### DISCUSSION TITLE
How to add audio on WebRTC / HLS streams with Gstreamer pipeline?

### DISCUSSION THREAD

**Discussion by Topper69** [bluenviron/mediamtx#1437] [Questions] [answered]
Hi,

I have an rasp4 + cam v2 + usb mic connected.

Part of my configuration file : 
path:
  mystream:
    runOnDemand: gst-launch-1.0 rtspclientsink name=s location=rtsp://localhost:8554/mystream libcamerasrc ! video/x-raw, width=640, height=480, framerate=30/1 ! videoconvert ! videoscale ! clockoverlay time-format="%d/%m/%y %H:%M:%S" ! x264enc speed-preset=ultrafast bitrate=600 key-int-max=40 ! queue ! s.sink_0 alsasrc ! queue ! voaacenc ! s.sink_1
    runOnDemandRestart: yes

When the client request the feed with VLC (rtsp://localhost:8554/mystream) i have video+audio.
With the same feed, when i connect through HLS (https://localhost:8888/mystream) or WebRTC (https://localhost:8889/mystream) i have only the video. No audio.

Server logs: 
2023/01/31 17:10:14 INF rtsp-simple-server v0.21.2
2023/01/31 17:10:14 INF [RTSP] listener opened on :8554 (TCP), :8000 (UDP/RTP), :8001 (UDP/RTCP)
2023/01/31 17:10:14 INF [RTSPS] listener opened on :8322 (TCP)
2023/01/31 17:10:14 INF [RTMP] listener opened on :1935
2023/01/31 17:10:14 INF [HLS] listener opened on :8888
2023/01/31 17:10:14 INF [WebRTC] listener opened on :8889 (HTTP)
2023/01/31 17:10:26 INF [path mystream] runOnDemand command started
Setting pipeline to PAUSED ...
[3:24:31.762910955] [3715]  INFO Camera camera_manager.cpp:299 libcamera v0.0.3+40-9b860a66
[3:24:31.807982154] [3717]  WARN RPI raspberrypi.cpp:1308 Mismatch between Unicam and CamHelper for embedded data usage!
[3:24:31.809371351] [3717]  INFO RPI raspberrypi.cpp:1425 Registered camera /base/soc/i2c0mux/i2c@1/imx219@10 to Unicam device /dev/media3 and ISP device /dev/media1
Pipeline is live and does not need PREROLL ...
Progress: (open) Opening Stream
Progress: (connect) Connecting to rtsp://localhost:8554/mystream
Pipeline is PREROLLED ...
Prerolled, waiting for progress to finish...
[3:24:31.827779313] [3723]  INFO Camera camera.cpp:1028 configuring streams: (0) 640x480-NV21
[3:24:31.828256508] [3717]  INFO RPI raspberrypi.cpp:805 Sensor: /base/soc/i2c0mux/i2c@1/imx219@10 - Selected sensor format: 640x480-SBGGR10_1X10 - Selected unicam format: 640x480-pBAA
Redistribute latency...
Progress: (open) Retrieving server options
2023/01/31 17:10:26 INF [RTSP] [conn [::1]:43888] opened
Redistribute latency...
Progress: (open) Opened Stream
Setting pipeline to PLAYING ...
New clock: GstAudioSrcClock
Progress: (request) Sending RECORD request
Progress: (record) Sending server stream info
2023/01/31 17:10:26 INF [RTSP] [session 023cca8c] created by [::1]:43888
Progress: (request) SETUP stream 1
Progress: (request) SETUP stream 0
Progress: (record) Starting recording
2023/01/31 17:10:28 INF [RTSP] [session 023cca8c] is publishing to path 'mystream', with UDP, 2 tracks (Generic, H264)
2023/01/31 17:10:28 INF [WebRTC] [conn 192.168.27.65:41336] opened
Redistribute latency...
WARNING: from element /GstPipeline:pipeline0/GstAlsaSrc:alsasrc0: Can't record audio fast enough
Additional debug info:
../gst-libs/gst/audio/gstaudiobasesrc.c(841): gst_audio_base_src_create (): /GstPipeline:pipeline0/GstAlsaSrc:alsasrc0:
Dropped 112896 samples. This is most likely because downstream can't keep up and is consuming samples too slowly.
2023/01/31 17:10:29 INF [WebRTC] [conn 192.168.27.65:41336] peer connection established, local candidate: host/udp/192.168.0.115/41018, remote candidate: prflx/udp/192.168.27.65/50028
2023/01/31 17:10:29 INF [WebRTC] [conn 192.168.27.65:41336] is reading from path 'mystream', 1 track (H264)
^C2023/01/31 17:10:40 INF shutting down gracefully

Many thanks for the help and btw very great app!

**Comment by Topper69** [comment]
My gsStreamer pipeline (working with some warning of pipeline0/GstAlsaSrc:alsasrc0: Can't record audio fast enough) : 
gst-launch-1.0 rtspclientsink name=s location=rtsp://localhost:8554/mystream libcamerasrc ! video/x-raw, width=640, height=480, framerate=30/1 ! videoconvert ! videoscale ! clockoverlay time-format="%d/%m/%y %H:%M:%S" ! x264enc speed-preset=ultrafast bitrate=600 key-int-max=40 ! queue ! s.sink_0 alsasrc ! queue ! voaacenc ! s.sink_1

**Comment by aler9** [comment, marked as answer]
Hello, use opusenc:

```
gst-launch-1.0 rtspclientsink name=s location=rtsp://localhost:8554/mystream libcamerasrc ! video/x-raw, width=640, height=480, framerate=30/1 ! videoconvert ! videoscale ! clockoverlay time-format="%d/%m/%y %H:%M:%S" ! x264enc speed-preset=ultrafast bitrate=600 key-int-max=40 ! queue ! s.sink_0 alsasrc ! queue ! audioconvert ! opusenc ! s.sink_1
```

**Reply by Topper69** [reply]
Many thanks!

Works like a charm.

**Comment by tux2linux** [comment]
I have a similar problem and opened an issue here: https://github.com/bluenviron/mediamtx/discussions/1970

[a comment by an automated account is not shown]
